fabric-particles-v1/src/main/java/net/fabricmc/fabric/impl/particles/FabricParticleManager.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.impl.particles;
+
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.Reader;
+import java.util.List;
+import java.util.Random;
+import java.util.stream.Collectors;
+
+import javax.annotation.Nullable;
+
+import com.google.common.base.Charsets;
+import com.google.common.collect.ImmutableList;
+
+import it.unimi.dsi.fastutil.ints.Int2ObjectMap;
+import it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap;
+import net.fabricmc.fabric.api.particles.FabricSpriteProvider;
+import net.minecraft.client.particle.ParticleTextureData;
+import net.minecraft.client.texture.Sprite;
+import net.minecraft.client.texture.SpriteAtlasTexture;
+import net.minecraft.resource.ResourceManager;
+import net.minecraft.util.Identifier;
+import net.minecraft.util.JsonHelper;
+
+public final class FabricParticleManager {
+	private final VanillaParticleManager manager;
+
+	private final Int2ObjectMap<FabricSpriteProviderImpl> providers = new Int2ObjectOpenHashMap<>();
+
+	public FabricParticleManager(VanillaParticleManager manager) {
+		this.manager = manager;
+	}
+
+	public void injectValues() {
+		manager.getFactories().putAll(ParticleFactoryRegistryImpl.INSTANCE.factories);
+		ParticleFactoryRegistryImpl.INSTANCE.constructors.forEach((id, factory) -> {
+			FabricSpriteProviderImpl provider = new FabricSpriteProviderImpl();
+
+			providers.put((int)id, provider);
+			manager.getFactories().put((int)id, factory.create(provider));
+		});
+	}
+
+	private FabricSpriteProviderImpl getProvider(Identifier id) {
+		if (!ParticleFactoryRegistryImpl.INSTANCE.constructorsIdsMap.containsKey(id)) {
+			return null;
+		}
+
+		return providers.get((int)ParticleFactoryRegistryImpl.INSTANCE.constructorsIdsMap.get(id));
+	}
+
+	public boolean loadParticle(ResourceManager manager, Identifier id) {
+		FabricSpriteProviderImpl provider = getProvider(id);
+
+		if (provider == null) {
+			return false; // preserve vanilla behaviour (i don't got dis)
+		}
+
+		Identifier file = new Identifier(id.getNamespace(), "particles/" + id.getPath() + ".json");
+
+		try (Reader reader = new InputStreamReader(manager.getResource(file).getInputStream(), Charsets.UTF_8)) {
```

Comment:
`StandardCharsets.UTF_8`

Reply:
This was taken from Mojang's code, so Charsets is what they were using.

Reply:
Don't use "mojang code" as an excuse. We need to updated code instead of keeping legacy code.

Reply:
Well what's the difference? They're the same constant, and what I'm saying is if Mojang is using it, why can't we? It's going to be part of the environment regardless.